snapshot.go

```diff
@@ -77,3 +78,113 @@ func (c *Client) UpdateSnapshot(snapshotID string, request SnapshotProperties) (
+// with logical AND.
+func (c *Client) ListSnapshotsWithSelector(selectors ...SnapshotSelector) ([]Snapshot, error) {
+	url := snapshotsPath()
+	ret := &Snapshots{}
+	err := c.Get(url, ret, http.StatusOK)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []Snapshot
+outerLoop:
+	for _, snapshot := range ret.Items {
+		for _, selector := range selectors {
+			if !selector(&snapshot) {
+				continue outerLoop
+			}
+		}
+		result = append(result, snapshot)
+	}
+	return result, nil
+}
+
+// SnapshotSelector is used to do client-side filtering of a list of Snapshots
+type SnapshotSelector func(*Snapshot) bool
+
+// SnapshotByState selects snapshots with the given state
+func SnapshotByState(state string) SnapshotSelector {
+	return func(snapshot *Snapshot) bool {
+		return snapshot.Metadata.State == state
+	}
+}
+
+// SnapshotByName selects snapshots with the given name
+func SnapshotByName(name string) SnapshotSelector {
+	return func(snapshot *Snapshot) bool {
+		return snapshot.Properties.Name == name
+	}
+}
+
+// SnapshotByDescription selects snapshots with the given description
+func SnapshotByDescription(description string) SnapshotSelector {
+	return func(snapshot *Snapshot) bool {
+		return snapshot.Properties.Description == description
+	}
+}
+
+// SelectExactSnapshot concatenates the provided selectors with logical AND.
+func SelectExactSnapshot(selectors ...SnapshotSelector) SnapshotSelector {
+	return func(snapshot *Snapshot) bool {
+		for _, selector := range selectors {
+			if !selector(snapshot) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+// SelectAnySnapshot concatenates the provided selectors with logical OR.
+func SelectAnySnapshot(selectors ...SnapshotSelector) SnapshotSelector {
+	return func(snapshot *Snapshot) bool {
+		for _, selector := range selectors {
+			if selector(snapshot) {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+// HasDeleteSnapshotInProgress tries to find an active delete snapshot request (QUEUED or RUNNING).
+func (c *Client) HasDeleteSnapshotInProgress(snapshotID string) (bool, error) {
+	f := NewRequestListFilter().WithUrl(snapshotPath(snapshotID)).WithMethod(http.MethodDelete)
+	result, err := c.ListRequestsWithFilter(f.Clone().WithRequestStatus(RequestStatusQueued))
+	if err != nil {
+		return false, err
+	}
+	if len(result.Items) > 0 {
+		return true, nil
+	}
+	result, err = c.ListRequestsWithFilter(f.Clone().WithRequestStatus(RequestStatusRunning))
+	if err != nil {
+		return false, err
+	}
+	return len(result.Items) > 0, nil
+}
+
+// IsSnapshotInUseError returns true if the given error indicates that a snapshot
+// is being referenced by a volume that is currently being created.
+func IsSnapshotInUseError(err error) bool {
+	apiErr, ok := err.(ApiError)
+	if !ok {
+		return false
+	}
+	return apiErr.HTTPStatus == http.StatusUnprocessableEntity &&
+		strings.Contains(apiErr.String(), "[VDC-18-0] An unexpected error occurred.")
```

Comment:
If there is an explicit error code in the error response body, we should use this. I think a method on ApiError that checks if ionos-error code is present in Messages would be fine. I see some places that can benefit from this (like Resource Exhausted Errors)